packages/core/src/workflows/handlers/sleep.ts

```diff
@@ -0,0 +1,255 @@
+    id: string;
+    duration?: number;
+    fn?: ExecuteFunction<any, any, any, any, any, DefaultEngineType>;
+  };
+  prevStep: StepFlowEntry;
+  prevOutput: any;
+  stepResults: Record<string, StepResult<any, any, any, any>>;
+  resume?: {
+    steps: string[];
+    stepResults: Record<string, StepResult<any, any, any, any>>;
+    resumePayload: any;
+    resumePath: number[];
+  };
+  executionContext: ExecutionContext;
+  emitter: Emitter;
+  abortController: AbortController;
+  requestContext: RequestContext;
+  writableStream?: WritableStream<ChunkType>;
+  tracingContext: TracingContext;
+}
+
+export async function executeSleep(engine: DefaultExecutionEngine, params: ExecuteSleepParams): Promise<void> {
+  const {
+    workflowId,
+    runId,
+    entry,
+    prevOutput,
+    stepResults,
+    emitter,
+    abortController,
+    requestContext,
+    executionContext,
+    writableStream,
+    tracingContext,
+  } = params;
+
+  let { duration, fn } = entry;
+
+  const sleepSpan = tracingContext.currentSpan?.createChildSpan({
+    type: SpanType.WORKFLOW_SLEEP,
+    name: `sleep: ${duration ? `${duration}ms` : 'dynamic'}`,
+    attributes: {
+      durationMs: duration,
+      sleepType: fn ? 'dynamic' : 'fixed',
+    },
+    tracingPolicy: engine.options?.tracingPolicy,
+  });
+
+  if (fn) {
+    const stepCallId = randomUUID();
+    duration = await engine.wrapDurableOperation(`workflow.${workflowId}.sleep.${entry.id}`, async () => {
+      return fn({
+        runId,
+        workflowId,
+        mastra: engine.mastra!,
+        requestContext,
+        inputData: prevOutput,
+        state: executionContext.state,
+        setState: (state: any) => {
+          executionContext.state = state;
+        },
+        retryCount: -1,
+        tracingContext: {
+          currentSpan: sleepSpan,
+        },
+        getInitData: () => stepResults?.input as any,
+        getStepResult: getStepResult.bind(null, stepResults),
+        // TODO: this function shouldn't have suspend probably?
+        suspend: async (_suspendPayload: any): Promise<any> => {},
+        bail: () => {},
+        abort: () => {
+          abortController?.abort();
+        },
+        [EMITTER_SYMBOL]: emitter,
+        [STREAM_FORMAT_SYMBOL]: executionContext.format,
+        engine: engine.getEngineContext(),
+        abortSignal: abortController?.signal,
+        writer: new ToolStream(
+          {
+            prefix: 'workflow-step',
+            callId: stepCallId,
+            name: 'sleep',
+            runId,
+          },
+          writableStream,
+        ),
+      });
+    });
+
+    // Update sleep span with dynamic duration
+    sleepSpan?.update({
+      attributes: {
+        durationMs: duration,
+      },
+    });
+  }
+
+  try {
+    await engine.executeSleepDuration(!duration || duration < 0 ? 0 : duration, entry.id, workflowId);
+    sleepSpan?.end();
+  } catch (e) {
+    sleepSpan?.error({ error: e as Error });
+    throw e;
+  }
```

Comment:
Can that catch statement ever be reached for the default engine? It looks like executeSleepDuration never rejects, so we never throw here. It was the same functionality before though, so just curious how that works.